File: sdks/dart_client/lib/src/data/query.dart
/// A PostgREST query builder — the data layer for Supabase and Neon.
///
/// Chain filters and modifiers, then `await` the builder (or call [execute]):
///
/// ```dart
/// final rows = await client
///     .from('todos')
///     .select('id, title, done')
///     .eq('user_id', userId)
///     .order('created_at', ascending: false)
///     .limit(20);
/// ```
///
/// Writes:
/// ```dart
/// await client.from('todos').insert({'title': 'Ship SDK'});
/// await client.from('todos').update({'done': true}).eq('id', id);
/// await client.from('todos').delete().eq('id', id);
/// ```
library;

import 'dart:async';

import '../http.dart';

/// Built by [ForgefyClient.from]. Each terminal operation (`select`, `insert`,
/// `update`, `delete`, `upsert`) sets the HTTP verb; filters and modifiers
/// append PostgREST query parameters.
///
/// Implements `Future` so it can be awaited directly. The request fires when
/// the future is first awaited/`.then`-ed, exactly once.
class ForgefyQuery implements Future<Object?> {
  ForgefyQuery(this._http, this._restUrl, this._table);

  final ForgefyHttp _http;
  final String _restUrl;
  final String _table;

  String _method = 'GET';
  Object? _body;
  final List<String> _params = [];
  final Map<String, String> _headers = {};
  bool _single = false;

  Future<Object?>? _pending;

  // ── Terminal verbs ─────────────────────────────────────────────────────────

  /// Read rows. [columns] is a PostgREST select list, e.g. `'id, title'` or
  /// `'*, author(name)'` for an embedded resource.
  ForgefyQuery select([String columns = '*']) {
    _method = 'GET';
    _params.add('select=${Uri.encodeQueryComponent(columns)}');
    return this;
  }

  /// Insert one row (a `Map`) or many (a `List<Map>`). Returns the inserted
  /// rows.
  ForgefyQuery insert(Object values) {
    _method = 'POST';
    _body = values;
    _headers['Prefer'] = 'return=representation';
    return this;
  }

  /// Insert-or-update on conflict. [onConflict] names the unique column(s) that
  /// define a duplicate; omit to use the table's primary key.
  ForgefyQuery upsert(Object values, {String? onConflict}) {
    _method = 'POST';
    _body = values;
    _headers['Prefer'] = 'return=representation,resolution=merge-duplicates';
    if (onConflict != null) {
      _params.add('on_conflict=${Uri.encodeQueryComponent(onConflict)}');
    }
    return this;
  }

  /// Patch rows matching the filters. Combine with `.eq(...)` etc. — PostgREST
  /// refuses an unfiltered update.
  ForgefyQuery update(Map<String, dynamic> values) {
    _method = 'PATCH';
    _body = values;
    _headers['Prefer'] = 'return=representation';
    return this;
  }

  /// Delete rows matching the filters. Combine with a filter — PostgREST
  /// refuses an unfiltered delete.
  ForgefyQuery delete() {
    _method = 'DELETE';
    _headers['Prefer'] = 'return=representation';
    return this;
  }

  // ── Filters ────────────────────────────────────────────────────────────────

  ForgefyQuery eq(String column, Object value) => _filter(column, 'eq', value);
  ForgefyQuery neq(String column, Object value) => _filter(column, 'neq', value);
  ForgefyQuery gt(String column, Object value) => _filter(column, 'gt', value);
  ForgefyQuery gte(String column, Object value) => _filter(column, 'gte', value);
  ForgefyQuery lt(String column, Object value) => _filter(column, 'lt', value);
  ForgefyQuery lte(String column, Object value) => _filter(column, 'lte', value);

  /// SQL `LIKE` (case-sensitive). Use `%` as the wildcard.
  ForgefyQuery like(String column, String pattern) =>
      _filter(column, 'like', pattern);

  /// SQL `ILIKE` (case-insensitive).
  ForgefyQuery ilike(String column, String pattern) =>
      _filter(column, 'ilike', pattern);

  /// `column IS value`, for `null` / `true` / `false`.
  ForgefyQuery isFilter(String column, Object? value) =>
      _filter(column, 'is', value ?? 'null');

  /// `column IN (values)`.
  ForgefyQuery inFilter(String column, List<Object> values) =>
      _filter(column, 'in', '(${values.join(',')})');

  // ── Modifiers ──────────────────────────────────────────────────────────────

  ForgefyQuery order(String column, {bool ascending = true}) {
    _params.add(
      'order=${Uri.encodeQueryComponent('$column.${ascending ? 'asc' : 'desc'}')}',
    );
    return this;
  }

  ForgefyQuery limit(int count) {
    _params.add('limit=$count');
    return this;
  }

  /// Zero-based inclusive row window (like SQL `OFFSET`/`LIMIT`).
  ForgefyQuery range(int from, int to) {
    _params.add('offset=$from');
    _params.add('limit=${to - from + 1}');
    return this;
  }

  /// Return a single object instead of a list, and 406 if not exactly one row.
  ForgefyQuery single() {
    _single = true;
    _headers['Accept'] = 'application/vnd.pgrst.object+json';
    return this;
  }

  // ── Execution ──────────────────────────────────────────────────────────────

  /// Fire the request (idempotent — repeated awaits reuse the same future).
  Future<Object?> execute() => _pending ??= _run();

  Future<Object?> _run() async {
    final query = _params.isEmpty ? '' : '?${_params.join('&')}';
    final res = await _http.send(
      _method,
      '$_restUrl/$_table$query',
      body: _body,
      headers: _headers,
      // GET is safe to retry; writes carry no idempotency key, so are not.
      retryOn5xx: _method == 'GET',
    );
    // With `Prefer: return=representation` PostgREST replies with a list even
    // for single-row writes; `.single()` unwraps it.
    if (_single && res.data is List<dynamic>) {
      final list = res.data as List<dynamic>;
      return list.isEmpty ? null : list.first;
    }
    return res.data;
  }

  ForgefyQuery _filter(String column, String op, Object value) {
    _params.add(
      '${Uri.encodeQueryComponent(column)}=${Uri.encodeQueryComponent('$op.$value')}',
    );
    return this;
  }

  // ── Future delegation ──────────────────────────────────────────────────────

  @override
  Future<R> then<R>(FutureOr<R> Function(Object? value) onValue,
          {Function? onError}) =>
      execute().then(onValue, onError: onError);

  @override
  Future<Object?> catchError(Function onError,
          {bool Function(Object error)? test}) =>
      execute().catchError(onError, test: test);

  @override
  Future<Object?> whenComplete(FutureOr<void> Function() action) =>
      execute().whenComplete(action);

  @override
  Stream<Object?> asStream() => execute().asStream();

  @override
  Future<Object?> timeout(Duration timeLimit,
          {FutureOr<Object?> Function()? onTimeout}) =>
      execute().timeout(timeLimit, onTimeout: onTimeout);
}

File: sdks/typescript_client/src/data/query.ts
/**
 * A PostgREST query builder — the data layer for Supabase and Neon.
 *
 * Chain filters and modifiers, then `await` the builder (it's a thenable) or
 * call {@link execute}:
 *
 * ```ts
 * const rows = await client
 *   .from<Todo>("todos")
 *   .select("id, title, done")
 *   .eq("user_id", userId)
 *   .order("created_at", { ascending: false })
 *   .limit(20);
 *
 * await client.from("todos").insert({ title: "Ship SDK" });
 * const row = await client.from<Todo>("todos").insert(t).single();
 * ```
 */

import type { ForgefyHttp } from "../http.js";

type Row = Record<string, unknown>;

/**
 * `Result` defaults to `T[]`; {@link single} re-types it to `T | null`. The
 * request fires when the builder is first awaited/`.then`-ed, exactly once.
 */
export class ForgefyQuery<T extends Row = Row, Result = T[]> implements PromiseLike<Result> {
  private method = "GET";
  private body: unknown;
  private readonly params: string[] = [];
  private readonly headers: Record<string, string> = {};
  private singleRow = false;
  private pending: Promise<Result> | undefined;

  constructor(
    private readonly http: ForgefyHttp,
    private readonly restUrl: string,
    private readonly table: string,
  ) {}

  // ── Terminal verbs ─────────────────────────────────────────────────────────

  /** Read rows. `columns` is a PostgREST select list, e.g. "id, title" or "*, author(name)". */
  select(columns = "*"): this {
    this.method = "GET";
    this.params.push(`select=${encodeURIComponent(columns)}`);
    return this;
  }

  /** Insert one row or an array of rows. Returns the inserted rows. */
  insert(values: Row | Row[]): this {
    this.method = "POST";
    this.body = values;
    this.headers.Prefer = "return=representation";
    return this;
  }

  /** Insert-or-update on conflict. `onConflict` names the unique column(s). */
  upsert(values: Row | Row[], opts: { onConflict?: string } = {}): this {
    this.method = "POST";
    this.body = values;
    this.headers.Prefer = "return=representation,resolution=merge-duplicates";
    if (opts.onConflict) this.params.push(`on_conflict=${encodeURIComponent(opts.onConflict)}`);
    return this;
  }

  /** Patch rows matching the filters (PostgREST refuses an unfiltered update). */
  update(values: Row): this {
    this.method = "PATCH";
    this.body = values;
    this.headers.Prefer = "return=representation";
    return this;
  }

  /** Delete rows matching the filters (PostgREST refuses an unfiltered delete). */
  delete(): this {
    this.method = "DELETE";
    this.headers.Prefer = "return=representation";
    return this;
  }

  // ── Filters ────────────────────────────────────────────────────────────────

  eq(column: string, value: unknown): this {
    return this.filter(column, "eq", value);
  }
  neq(column: string, value: unknown): this {
    return this.filter(column, "neq", value);
  }
  gt(column: string, value: unknown): this {
    return this.filter(column, "gt", value);
  }
  gte(column: string, value: unknown): this {
    return this.filter(column, "gte", value);
  }
  lt(column: string, value: unknown): this {
    return this.filter(column, "lt", value);
  }
  lte(column: string, value: unknown): this {
    return this.filter(column, "lte", value);
  }
  /** SQL LIKE (case-sensitive); use `%` as the wildcard. */
  like(column: string, pattern: string): this {
    return this.filter(column, "like", pattern);
  }
  /** SQL ILIKE (case-insensitive). */
  ilike(column: string, pattern: string): this {
    return this.filter(column, "ilike", pattern);
  }
  /** `column IS value`, for null / true / false. */
  is(column: string, value: null | boolean): this {
    return this.filter(column, "is", value === null ? "null" : value);
  }
  /** `column IN (values)`. */
  in(column: string, values: unknown[]): this {
    return this.filter(column, "in", `(${values.join(",")})`);
  }

  // ── Modifiers ──────────────────────────────────────────────────────────────

  order(column: string, opts: { ascending?: boolean } = {}): this {
    const dir = opts.ascending === false ? "desc" : "asc";
    this.params.push(`order=${encodeURIComponent(`${column}.${dir}`)}`);
    return this;
  }

  limit(count: number): this {
    this.params.push(`limit=${count}`);
    return this;
  }

  /** Zero-based inclusive row window (like SQL OFFSET/LIMIT). */
  range(from: number, to: number): this {
    this.params.push(`offset=${from}`, `limit=${to - from + 1}`);
    return this;
  }

  /** Return a single object instead of a list, and error if not exactly one row. */
  single(): ForgefyQuery<T, T | null> {
    this.singleRow = true;
    this.headers.Accept = "application/vnd.pgrst.object+json";
    return this as unknown as ForgefyQuery<T, T | null>;
  }

  // ── Execution ──────────────────────────────────────────────────────────────

  /** Fire the request (idempotent — repeated awaits reuse the same promise). */
  execute(): Promise<Result> {
    return (this.pending ??= this.run());
  }

  then<TResult1 = Result, TResult2 = never>(
    onfulfilled?: ((value: Result) => TResult1 | PromiseLike<TResult1>) | null,
    onrejected?: ((reason: unknown) => TResult2 | PromiseLike<TResult2>) | null,
  ): Promise<TResult1 | TResult2> {
    return this.execute().then(onfulfilled, onrejected);
  }

  catch<TResult = never>(
    onrejected?: ((reason: unknown) => TResult | PromiseLike<TResult>) | null,
  ): Promise<Result | TResult> {
    return this.execute().catch(onrejected);
  }

  private async run(): Promise<Result> {
    const query = this.params.length === 0 ? "" : `?${this.params.join("&")}`;
    const res = await this.http.send(this.method, `${this.restUrl}/${this.table}${query}`, {
      body: this.body,
      headers: this.headers,
      // GET is safe to retry; writes carry no idempotency key, so are not.
      retryOn5xx: this.method === "GET",
    });
    // With `Prefer: return=representation` PostgREST replies with a list even
    // for single-row writes; `.single()` unwraps it.
    if (this.singleRow && Array.isArray(res.data)) {
      return (res.data.length === 0 ? null : res.data[0]) as Result;
    }
    return res.data as Result;
  }

  private filter(column: string, op: string, value: unknown): this {
    this.params.push(`${encodeURIComponent(column)}=${encodeURIComponent(`${op}.${String(value)}`)}`);
    return this;
  }
}
